src/approaches/grammar_search_invention_approach.py

```diff
@@ -667,6 +670,101 @@ def _evaluate_with_operators(self,
+                                 strips_ops: List[STRIPSOperator],
+                                 option_specs: List[OptionSpec]) -> float:
+        del segments  # unused
+        score = 0.0
+        seen_demos = 0
+        for traj, atoms_sequence in pruned_atom_data:
+            if seen_demos >= CFG.grammar_search_max_demos:
+                break
+            if not traj.is_demo:
+                continue
+            seen_demos += 1
+            init_atoms = atoms_sequence[0]
+            goal = traj.goal
+            # Ground everything once per demo.
+            objects = set(traj.states[0])
+            ground_nsrts, reachable_atoms = task_plan_grounding(
+                init_atoms, objects, strips_ops, option_specs)
+            heuristic = utils.create_task_planning_heuristic(
+                CFG.task_planning_heuristic, init_atoms, goal, ground_nsrts,
+                candidate_predicates | self._initial_predicates, objects)
+            # The expected number of nodes that need to be created before a
+            # refinable skeleton is found.
+            expected_num_nodes = 0.0
+            # Keep track of the probability that a refinable skeleton has still
+            # not been found, updated after each new goal-reaching skeleton is
+            # considered.
+            refinable_skeleton_not_found_prob = 1.0
+            generator = task_plan(init_atoms, goal, ground_nsrts,
+                                  reachable_atoms, heuristic, CFG.seed,
+                                  CFG.grammar_search_task_planning_timeout)
+            try:
+                for (_, plan_atoms_sequence, metrics) in generator:
+                    assert goal.issubset(plan_atoms_sequence[-1])
+                    # Estimate the probability that this skeleton is refinable.
+                    refinement_prob = self._get_refinement_prob(
+                        atoms_sequence, plan_atoms_sequence)
+                    # Get the number of nodes that have been created so far.
+                    num_nodes = metrics["num_nodes_created"]
+                    # This contribution to the expected number of nodes is for
+                    # the event that the current skeleton is refinable, but no
+                    # previous skeleton has been refinable.
+                    p = refinable_skeleton_not_found_prob * refinement_prob
+                    expected_num_nodes += p * num_nodes
+                    # Update the probability that no skeleton yet is refinable.
+                    refinable_skeleton_not_found_prob *= (1 - refinement_prob)
+            except (ApproachTimeout, ApproachFailure):
+                # Note if we failed to find any skeleton, the next lines add
+                # the upper bound with refinable_skeleton_not_found_prob = 1.0,
+                # so no special action is required.
+                pass
+            # After exhausting the skeleton budget or timeout, we use this
```

Comment:
the simpler thing i mentioned above is equivalent to a pessimistic assumption that we are screwed after exhausting the budget?

Reply:
well I'm not sure how that would work. what would we add to the expected node count? what probability would we use?

Reply:
that's true, we'd need to score based on the tuple (refinable_skeleton_not_found_prob, expected_nodes), where the latter tiebreaks for the former